src/nodes.rs

```diff
@@ -0,0 +1,93 @@
+use crate::geoip_cache::GeoIPCache;
+use ziggurat_core_geoip::geoip::GeoInfo;
+
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+use spectre::graph::Graph;
+
+#[derive(Default, Clone, Serialize, Deserialize)]
+pub struct Node {
+    pub ip: String,
+    pub betweenness: f64,
+    pub closeness: f64,
+    pub column_position: u32,
+    pub column_size: u32,
+    pub connections: Vec<usize>,
+    pub geolocation: Option<GeoInfo>,
+}
+
+fn hash_geo_location(latitude: f64, longitude: f64) -> String {
+    // make unique every 0.2, so multiply by 5, convert to integer
+    let ilatitude: i32 = (latitude * 5.0).floor() as i32;
+    let ilongitude: i32 = (longitude * 5.0).floor() as i32;
+    format!("{ilatitude}:{ilongitude}")
+}
+
+// essentially, we sort the nodes into groups at (nearly) same geo-location
+// we use 0.2 degrees for epsilon in both axes.
+// hash gets created from a string created by two numbers
+// increment each time the same location is found
+pub fn set_column_positions(nodes: &mut Vec<Node>) -> HashMap<String, u32> {
+    let mut column_stats: HashMap<String, u32> = HashMap::new();
+    for node in nodes {
+        if let Some(geoinfo) = &node.geolocation {
+            if let Some(latitude) = geoinfo.latitude {
+                if let Some(longitude) = geoinfo.longitude {
+                    let geostr = hash_geo_location(latitude, longitude);
+                    column_stats
+                        .entry(geostr.clone())
+                        .and_modify(|count| *count += 1)
+                        .or_insert(1);
+                    node.column_position = column_stats[&geostr];
+                }
+            }
+        }
+    }
+    column_stats
+}
+
+// do the lookup again, and set node's corresponding final column size
+// this field is the same for all nodes in a given column
+pub fn set_column_sizes(nodes: &mut Vec<Node>, column_stats: &mut HashMap<String, u32>) {
+    for node in nodes {
+        if let Some(geoinfo) = &node.geolocation {
+            if let Some(latitude) = geoinfo.latitude {
+                if let Some(longitude) = geoinfo.longitude {
```

Comment:
I'll look into this after IPS.
Online providers guarantee to give us both latitude and longitude (https://github.com/grantshandy/ipgeolocate/blob/main/src/lib.rs) of specific location.
Offline provider read those two values separately and give us `Option<>` for each of them:
https://github.com/marirs/rust-ip2location/blob/e9bbb0c2cba3ce197fdc8f5000e1cc3bc9c297a0/src/ip2location/db.rs#L277
which makes it potentially possible to have only one of the values.
Will think about the API to handle such case.

Reply:
> which makes it potentially possible to have only one of the values.

that shouldn't happen unless it's a bug, no?

I just checked the example ip2location file, and the number of longitude occurrences is the same as the number of latitude occurrences:
```zsh
➜  crunchy git:(main) ✗ grep -o longitude geoip-cache.json | wc -l
    2531
➜  crunchy git:(main) ✗ grep -o latitude geoip-cache.json | wc -l
    2531
```

Reply:
Yes, the question is if the database contains also both params for each IP. Cache is only subset from database ;)  
But in general, there ought to be returned both lat and long factors - that's why I wrote about only potential possibility and maybe we can just assume default value if one of them is not returned (or even both) as I don't expect it will be often case where location would be unknown. If exact location is not known those bases usually use ISP HQ location (from cache observation ;)) so will change it soon.

Reply:
> done (for now) -- implemented @Rqnsom's suggestion

we'll see if we this leads to a panic eventually - then we can discuss this again